README
Music Tools - command line utilities for music composition and analysis

To install this module, use CPAN, for example:

  cpanm App::MusicTools

For the current source tree, see:

http://github.com/thrig/App-MusicTools

which can be built and installed via:

  perl Makefile.PL
  make
  make test
  make install

However, if installing from the github directory, the current versions of
various other music modules may also need be installed from github. This can be
done with the script included in this distribution thusly:

  sh App-MusicTools/install-music-modules

DEPENDENCIES

Some knowledge of lilypond and music theory may well benefit the user. See the
Makefile.PL for module dependencies.

COPYRIGHT AND LICENCE

Copyright (C) 2012-2015 Jeremy Mates

This module is free software; you can redistribute it and/or modify it
under the Artistic License (2.0).


Changes
 * 1.31 - relicense under BSD. TODO
 * 1.30 - atonal-util freq2pitch shows delta from expected frequence of
          the resulting pitch. canonical adds 'map' flag.
 * 1.29 - canonical multicolumn read on stdin, Test::Cmd test rewrite, etc.
 * 1.28 - atonal-util notes2time --fraction=..., basic shows common
   names for certain pitch sets, fix a dep.
 * 1.27 - atonal-util adjacent_interval_content + --sd improvements.
 * 1.25 - atonal-util time2notes so can reverse the notes2time ms.
 * 1.24 - mandate POD tests, in contrary motion to the fad of skimping on such.
 * 1.22 - support different languages with ly-fu mostly so can use with
   Music::PitchNum::German output. Also, ensure script headers reference the
   module that installs them.
 * 1.19 - reworked ly-fu: add --show-code option, better output, etc.

Otherwise, see the git repository.


t/Util.pm
# Shamelessly stolen and adapted from:
# http://api.metacpan.org/source/PETDANCE/ack-1.96/t/Util.pm

# capture stderr output into this file
my $catcherr_file = 'stderr.log';

sub is_win32 {
  return $^O =~ /Win32/;
}

sub build_command_line {
  my @args = @_;

  if ( is_win32() ) {
    for (@args) {
      s/(\\+)$/$1$1/;    # Double all trailing backslashes
      s/"/\\"/g;         # Backslash all quotes
      $_ = qq{"$_"};
    }
  } else {
    @args = map { quotemeta $_ } @args;
  }

  return "$^X ./capture-stderr $catcherr_file @args";
}

sub run_util {
  my @args = @_;

  my ( $stdout, $stderr ) = run_cmd_with_stderr(@args);

  if ($TODO) {
    fail(q{Automatically fail stderr check for TODO tests.});
  } else {
    is( scalar @{$stderr}, 0, "Should have no output to stderr: @args" )
      or diag( join( "\n", "STDERR:", @{$stderr} ) );
  }

  return wantarray ? @{$stdout} : join( "\n", @{$stdout} );
}

{    # scope for $cmd_return_code;

  # capture returncode
  our $cmd_return_code;

  # run the given command, assuming that the command was created with
  # build_command_line (and thus writes its STDERR to $catcherr_file).
  #
  # sets $cmd_return_code and unlinks the $catcherr_file
  #
  # returns chomped STDOUT and STDERR as two array refs
  sub run_cmd {
    my $cmd = shift;

    #diag( "Running command: $cmd" );

    my @stdout = `$cmd`;
    my ( $sig, $core, $rc ) = ( ( $? & 127 ), ( $? & 128 ), ( $? >> 8 ) );
    $ack_return_code = $rc;
    ## XXX what do do with $core or $sig?

    my @stderr;
    open( my $fh, '<', $catcherr_file ) or die $!;
    while (<$fh>) {
      push( @stderr, $_ );
    }
    close $fh or die $!;
    unlink $catcherr_file;

    chomp @stdout;
    chomp @stderr;

    return ( \@stdout, \@stderr );
  }

  sub get_rc {
    return $cmd_return_code;
  }

}    # scope for $cmd_return_code

sub run_cmd_with_stderr {
  my @args = @_;

  my @stdout;
  my @stderr;

  my $cmd = build_command_line(@args);

  return run_cmd($cmd);
}

# pipe into cmd and return STDOUT and STDERR as array refs
sub pipe_into_cmd_with_stderr {
  my $input = shift;
  my @args  = @_;

  my $cmd = build_command_line(@args);
  $cmd = "$^X -pe1 $input | $cmd";

  my ( $stdout, $stderr ) = run_cmd($cmd);
  return ( $stdout, $stderr );
}

# pipe into cmd and return STDOUT as array, for arguments see
# pipe_into_cmd_with_stderr
sub pipe_into_cmd {
  my ( $stdout, $stderr ) = pipe_into_cmd_with_stderr(@_);
  return @$stdout;
}

1;


lib/App/MusicTools.pm
# -*- Perl -*-
#
# Of more interest may be the scripts listed as EXE_FILES in Makefile.PL.

package App::MusicTools;
use 5.014000;
our $VERSION = '1.31';
1;

__END__

=head1 NAME

App-MusicTools - command line utilities for music composition and analysis

=head1 SYNOPSIS

  $ atonal-util --ly findin --pitchset=5-25 c e g
  Ti(0)   c,e,g,a,ais
  T(4)    c,e,g,fis,a

  $ canonical --contrary --retrograde modal c f g e c
  c g, f, a, c

  $ canonical modal --input=dorian --output=locrian d e f g a b c
  d dis f g gis ais c

  $ harmonic-fit f e ees d g aes ges b c a bes des
  ...

  $ export MIDI_EDITOR=timidity
  $ ly-fu --instrument=banjo c des ees c des bes c aes

  $ scalemogrifier --mode=mixolydian --transpose=fis
  fis gis ais b cis dis e fis'

  $ varionator '(I I6) (II IV II6 IV6) (V V7 III) I'
  ...

  $ vov --outputtmpl='<%{chord}> \t% %{vov}' I IV/IV V7/IV IV V7 I
  <c e g> 	% I
  <b dis f> 	% IV/IV
  <c e g b> 	% V7/IV
  <f a c> 	% IV
  <g b d f> 	% V7
  <c e g> 	% I

And more!

=head1 DESCRIPTION

This distribution contains various command line utilities related to
music composition and analysis, brief examples of which are shown in the
L</"SYNOPSIS">, above. Run C<perldoc> on the command name for additional
documentation on how to use each tool, as several of the programs have
many, many more options than shown here.

The C<zsh-compdef> directory of this distribution includes ZSH
completion scripts for the command line utilities. Install these files
to a ZSH C<$fpath> directory, and follow the ZSH documentation on setting
up tab completion.

=head1 BUGS

Probably lots. Send patches, as there are probably lots between my bad
Perl and badder understanding of music theory.

=head1 SEE ALSO

=over 4

=item *

L<http://www.lilypond.org/> and most notably the Learning and Notation
manuals, as the tools make use of the default lilypond note letter
conventions (e.g. fis for F sharp, ees for E flat, etc.)

=item *

C<ly-fu> expects that a MIDI player and PDF viewer are available on the system.
Environment variables will likely need to be set or the source code tweaked as
necessary.

=item *

"The Structure of Atonal Music" by Allen Forte (ISBN 978-0-300-02120-2).
Though, note that there are two forms of pitch set construction; the one
created by Rahn is more common in computing circles.

=item *

B<Theory of Harmony> by Arnold Schoenberg (ISBN 978-0-520-26608-7).

=item *

L<http://en.wikipedia.org/wiki/Forte_number>

=item *

L<Music::AtonalUtil>, L<Music::Chord::Positions>,
L<Music::LilyPondUtil>, L<Music::Tension> and others.

=back

=head1 AUTHOR

thrig - Jeremy Mates (cpan:JMATES) C<< <jmates at cpan.org> >>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2012-2016,2019 by Jeremy Mates

This program is distributed under the (Revised) BSD License:
L<http://www.opensource.org/licenses/BSD-3-Clause>

=cut


Makefile.PL
#!perl

# was at 5.010, but some Windows smoke box ignored the MIN_PERL_VERSION
# and then failed all over the 5.14 features in a *.t file.
use 5.014000;

use strict;
use warnings;
use ExtUtils::MakeMaker;

# perhaps stolen from mauke/Dir-Self
sub merge_key_into {
  my ( $href, $target, $source ) = @_;
  %{ $href->{$target} } = ( %{ $href->{$target} }, %{ delete $href->{$source} } );
}

my %params = (
  NAME         => 'App::MusicTools',
  AUTHOR       => 'Jeremy Mates <jmates@cpan.org>',
  ABSTRACT     => 'command line utilities for music composition and analysis',
  VERSION_FROM => 'lib/App/MusicTools.pm',

  EXE_FILES => [
    qw/atonal-util canonical harmonic-fit ly-fu scalemogrifier varionator vov/
  ],

  LICENSE          => 'artistic_2',
  MIN_PERL_VERSION => 5.14.0,

  CONFIGURE_REQUIRES => {
    'strict'              => 0,
    'ExtUtils::MakeMaker' => '6.48',
  },
  BUILD_REQUIRES => {},
  TEST_REQUIRES  => {
    'Test::Cmd'       => 0,
    'Test::Exception' => 0,
    'Test::Most'      => 0,
    # Mandate the POD tests, contrary to the dagolden no-pod-tests fad. I rank
    # documentation on par with the code, not something one can maybe remember
    # to maybe run the release testing for.
    'Pod::Coverage'       => 0.18,
    'Test::Pod::Coverage' => 1.08,
    'Test::Pod'           => 1.22,
    'lib'                 => 0,
  },
  # List everything due to certain linux vendors chopping up core perl, or
  # otherwise to detect and fail if being installed to a broken perl install.
  PREREQ_PM => {
    Carp                      => 0,
    'File::Spec'              => 0,
    'File::Temp'              => 0,
    'Getopt::Long'            => 0,
    'IPC::Run3'               => 0,
    'JSON'                    => 0,
    'List::Util'              => 0,
    'Music::AtonalUtil'       => 1.14,
    'Music::Canon'            => 2.00,
    'Music::Chord::Positions' => 0,
    'Music::LilyPondUtil'     => 0.10,
    'Music::Scala'            => 0.84,
    'Music::Scales'           => 0,
    'Music::Tempo'            => 0,
    'Music::Tension'          => 0.60,
    'Parse::Range'            => 0,
    'Scalar::Util'            => 0,
    'Text::Roman'             => 0,
    'Text::Wrap'              => 0,
  },

  META_MERGE => {
    'meta-spec' => { version => 2 },
    'resources' => {
      repository => {
        url  => 'git://github.com/thrig/App-MusicTools.git',
        web  => 'https://github.com/thrig/App-MusicTools',
        type => 'git',
      },
    },
  },

  dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
  clean => { FILES    => 'App-MusicTools-*' },
);

( my $mm_version = ExtUtils::MakeMaker->VERSION(
    $params{CONFIGURE_REQUIRES}{'ExtUtils::MakeMaker'}
  )
) =~ tr/_//d;

if ( $mm_version < 6.67_04 ) {
  # also stolen from mauke
  *ExtUtils::MM_Any::_add_requirements_to_meta_v1_4 =
    \&ExtUtils::MM_Any::_add_requirements_to_meta_v2;
}

if ( $mm_version < 6.63_03 ) {
  merge_key_into \%params, 'BUILD_REQUIRES', 'TEST_REQUIRES';
}

if ( $mm_version < 6.55_01 ) {
  merge_key_into \%params, 'CONFIGURE_REQUIRES', 'BUILD_REQUIRES';
}

if ( $mm_version < 6.51_03 ) {
  merge_key_into \%params, 'PREREQ_PM', 'CONFIGURE_REQUIRES';
}

WriteMakefile( ( MM->can('signature_target') ? ( SIGN => 1 ) : () ), %params );
